mindmeld/components/dialogue.py

```diff
@@ -689,6 +692,189 @@ def _get_dialogue_handler(self, dialogue_state):
+        self._retry_attempts = retry_attempts
+        self._local_form = None
+        self._slot_not_prompted = None
+
+    def _set_target_state(self, responder):
+        """Set target dialogue state to current flow"""
+        responder.params.target_dialogue_state = self._entrance_handler.__name__
+
+    def _exit_flow(self, responder):
+        """Exits this flow and clears the related parameter for re-usability"""
+        self._slot_not_prompted = None
+        self._local_form = None
+        responder.exit_flow()
+
+    def _extract_query_features(self, text):
+        """ Extracts Query object from the user input and converts it into
+        appropriate format for entity extraction.
+
+        Args:
+            text (Request.text): Text in the query
+
+        Returns:
+            Formatted Payload (tuple(query (Query Object), list of entities, entity index))
+        """
+        resource_loader = self._app.app_manager.nlp.resource_loader
+
+        """ Extracting the Query object associated with the user input here.
+
+        process_markup args:
+            text, query factory, query options (optional dict) {language, time_zone and time_stamp}
+
+        returns:
+            tuple(raw_text (str), Query object, list of entities)
+
+        The following returns the extracted Query object
+        """
+        query = process_markup(text, resource_loader.query_factory, {})[1]
+
+        return query
+
+    def _validate(self, text, entity_type, slot):
+        """Validates the user input based on the entity type and validation type.
+
+        Args:
+            text (Request.text): Text in the query
+            entity_type: Type of entity
+            slot: FormEntity object
+
+        Returns:
+            (bool): Boolean whether the user input fulfills the slot requirements
+        """
+
+        query = self._extract_query_features(text)
+
+        """
+        payload format for entity feature extractors:
+            tuple(query (Query Object), list of entities, entity index)
+
+        For slot filling, the user query will consist of the prompted entity.
+        Hence `entity = [query]` and `entity_index = 0`.
+
+        """
+        formatted_payload = (query, [query], 0)
+
+        if slot.default_eval:
+            if entity_type in DEFAULT_SYS_ENTITIES:
+                # system entity validation - checks for presence of required system entity
+                resources = {}
+                extracted_feature = dict(
+                    query_features.extract_sys_candidates([entity_type])(query, resources)
+                )
+            else:
+                # gazetteer validation
+                gazetteer = (
+                    {'gazetteers':
+                        {entity_type:
+                            self._app.app_manager.nlp.resource_loader.get_gazetteer(entity_type)
+                        }
+                    }
+                )
+                extracted_feature = (
```

Comment:
here extracted feature is a tuple, not a dict like other cases?

Reply:
It is a dict.